ORTmodule onnx input support

almaly-microsoft opened the pull request:
ORTmodule wraps pytorch model and executes training for it. However pytorch models are stored in ONNX format and in some cases (like quantization emulation) they needs to be retrained to get accuracy back.

This change is about adding that support. I updated one test to store pytorch model into ONNX format and then passed it into ORTModule. Training runs well.

almaly-microsoft committed:
Training support, with the issue on back prop parameters update

almaly-microsoft committed:
Trainig runs well for ONNX model

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:
```diff
@@ -81,29 +90,33 @@ def __init__(self, module):
 
         # flag to enable symbolic shape inference for dynamic shape inputs to improve performance
-        self._run_symbolic_shape_infer = True
+        self._run_symbolic_shape_infer = False
```
rollback

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:
```diff
@@ -81,29 +90,33 @@ def __init__(self, module):
         self.is_rocm_pytorch = (True if ((torch.version.hip is not None) and (ROCM_HOME is not None)) else False)
 
-        self._use_external_gpu_allocator = True
+        self._use_external_gpu_allocator = False
```
rollback

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/ortmodule.py:
```diff
@@ -40,44 +42,66 @@ class ORTModule(torch.nn.Module):
+                  Outputs are a concatenation of the loss function's output and the
+                  model's output.
+               - a combined ONNX model and loss function.
+            model_desc: Specify input/output shapes, types, and names.
```
remove

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/_io.py:
```diff
@@ -18,44 +18,41 @@ def __init__(self,
         self.keyword_names = keyword_names
 
     def __repr__(self) -> str:
         return f'''_InputInfo class:
```
rollback formatting

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/_training_manager.py:
```diff
@@ -10,6 +10,7 @@
 from onnxruntime.capi import _pybind_state as C
 from onnxruntime.capi.onnxruntime_inference_collection import get_ort_device_type
 
+import copy
```
check if needed

almaly-microsoft replied:
removed

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/_training_manager.py:
```diff
@@ -151,13 +158,10 @@ def backward(ctx, *grad_outputs):
-                del ctx.run_info.state
                 # Return input and initializer gradients
-                num_user_input_grads = len(self._input_info.require_grad_names)
+                num_user_input_grads = len(self._input_info.require_grad_names if self._input_info else self._graph_info.user_input_grad_names)
```
check if needed

almaly-microsoft replied:
seems not needed so far

almaly-microsoft commented on orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:
```diff
@@ -10,9 +10,10 @@
 import copy
 import io
-import inspect
+import inspect 
```
remove space

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:
```diff
@@ -41,15 +42,23 @@ def __init__(self, module):
+            self._onnx_model = module
+            # Model after inference optimization or gradient building.
+            self._optimized_onnx_model = module
+            self._onnx_model_parameters = onnx_model_parameters
```
maybe you could iterate over `module.graph.initializer` and fill in this without user interaction?

almaly-microsoft replied:
yep, thanks. No reason for execution manager to know about model parameters

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/ortmodule.py:
```diff
@@ -113,8 +137,12 @@ def get_buffer(self, target: str) -> torch.Tensor:
+                yield parameter
+        else:
+            """Override original method to delegate execution to the base module"""
+            yield from self._original_module.parameters(recurse=recurse)
```
you would have to override all `torch.nn.Module` APIs to prevent runtime errors

almaly-microsoft replied:
corrected, please check

almaly-microsoft committed:
Code clean up + addressing some comments

almaly-microsoft committed:
merge with master

almaly-microsoft committed:
after merge fixes

almaly-microsoft committed:
Fix for ORTModule interface when ONNX model is passed into it

almaly-microsoft committed:
Added tests call to cover added funcitonality

almaly-microsoft committed:
merge with master

baijumeswani committed:
Add module attribute to ORTModule to support HuggingFace Trainer save_model (#8088)

Commit:
Fix input schema extrator for ORTModule (#8098)

pengwa committed:
clean up builds for interop_torch (#8017)

* clean up builds for interop_torch

* add python dependency for executables

* disable onnxruntime_ENABLE_TRAINING_TORCH_INTEROP by default; enable it in ortmodule GPU training pipeline only

* disable training unrelated tests when torch interop is enabled

* simplify the python dependency.

* clean up and fix

souptc committed:
support pass in custom op registry for eager mode (#8087)

* support pass in custom op registry for eager mode

* fix the comments

SherlockNoMad committed:
Relax test tolerance to make CI more reliable (#8100)

BowenBao committed:
Add NGramRepeatBlock contrib op (#8078)

**Description**: 
Enforce no repetition of n-grams. Scores are set to `-inf` for tokens that form a repeated n-gram if added to the back of the input_ids.

**Motivation and Context**
Needed by transformer models in sequence generation algorithms (greedy search and beam search). This module has heavy impact on performance, and can be highly parallelized.

Commit:
Split Linux CPU CI pipeline (#8097)

chandru-r committed:
Added op_name to message when we are missing a kernel. (#8110)

* Added op_name to message when we are missing a kernel.

* Added domain and version

* Added missing ,

almaly-microsoft committed:
after merge with master fixes

duli2012 committed:
Adding Conv+Clip fusion (#8102)

chethanpk committed:
Added support for ReduceMean on DNNL EP for CPU and GPU (#7902)

* Added support for ReduceMean on DNNL EP for CPU and GPU

Signed-off-by: Chethan Palangotu Keshava <chethan.palangotu.keshava@intel.com>

* Added fix for a resnet model failure where it was failing to create dst shape for reducemean when it was part of a subgraph with other ops

Signed-off-by: Chethan Palangotu Keshava <chethan.palangotu.keshava@intel.com>

* Removing the DNNL EP from these unit tests. This is in anticipation of two changes:
- DNNL EP unit tests would be added in a different location later on, so addition of EP individually to these tests will not be necessary
- This was causing a memory leak fail in debug build. The bug is in the EP itself and not in the code added for reducemean. The fix for this is in the i/o handling overhaul which will be added later.

* Update reduction_ops_test.cc

Had accidentally deleted a new line. Making sure there are no unnecessary changes in this file

almaly-microsoft committed:
Fixes and rollbacks to align with master

chilo-ms committed:
Add TRT 7.1 Pipeline (#8073)

* Revert for testing TensorRT 7.1

* change to origianl googletest version

* change machine

* remove build arg

* change back machine

* revert back googletest version

* Make it ready to merge to master

* revert onnx-tensorrt to v7.1

* rename yml

* use [[ ]] in bash command

* add sudo

* add chmod

* add correct path

* change another way to revert onnx-tensorrt

* change docker image to manylinux build

hanbitmyths committed:
remove debug.keystore from repository due to a credential issue report (#8113)

Commit:
Col2im optimization by eliminating integer multiplications:

1. No padding branch performance is improved 8 times
2. Symmetric padding branch is generalized for asymmetric padding case (padding symmetry was not actually used) and further optimized by eliminating integer multiplications.

Commit:
Fix a compile error in InferenceTest.cs (#8119)

scxiao committed:
Update migraphx to rocm4.2 (#7994)

* update dockerfile for migraphx ep

* update to rocm4.2

* code cleanup

* fix error related to onnx unit tests

Commit:
Implement WINRT_IMPL_LoadLibraryW to avoid calling LoadLibraryW directly (#8065)

* Override load library w in cppwinrt

* Add comment

yufenglee committed:
initialize generated_value_names with graph input (#8085)

* initialize generated_value_names with graph input
* use set for following usage

guoyu-wang committed:
Fix 32bit Android java API crash (#8122)

* Fix 32bit Android java API crash

* fix code formating

guoyu-wang committed:
[Android] Output error message to android log instead of stderr (#8114)

* Output error message to android log instead of stderr

* Address CR comments, move macro to a helper function

* Address CR comments

* Fix ort minimal build break

almaly-microsoft committed:
Separated training and inference managers

almaly-microsoft committed:
Rollback training and inference managers for torch.nn.Model

Commit:
Delete some unused code in run_dockerbuild.sh and Enable Nuget CUDA tests (#8089)

1. Remove some unused code and simplify tools/ci_build/github/linux/run_dockerbuild.sh.
2. Enable Nuget CUDA tests. The original design was we could leverage Directory.Build.props and let cmake generate the required properties(USE_CUDA/...) there. However, in nuget packaging pipeline we test the package on a different host that doesn't run cmake command and doesn't have the auto-generated Directory.Build.props file.

edgchen1 committed:
[Mobile package] Update required operator config with additional ops for newer version of Wav2Vec 2. (#8123)

This is an update to https://github.com/microsoft/onnxruntime/pull/8079
The sample application motivating the original update changed to use an updated version of the model. Now, fewer ops are required. This change removes the previously added ops which are no longer needed.

skottmckay committed:
Add int64 as a required type to ConstantOfShape as it's used by the pytorch converter for Pad. (#8128)

It's also used pointlessly for torch.tensor.repeat (although that usage should always be able to be constant folded).

xi-liu-ds committed:
User dynamic axes in one step beam search output (#8092)

tracysh committed:
MLAS: enable SSE 4.1 path for x86 build (#8127)

Commit:
[js] fixing broadcast issues in pack mode (#8090)

* fixing broadcast issues in pack mode

* improved bcast logic for matmul

* removed TODO

* rebased from master

Commit:
Chang how numpy version is handled. (#8130)

Numpy has binary compatibility, which means "binaries compiled against a given version of NumPy will still run correctly with newer NumPy versions, but not with older versions." So, if an onnx runtime package was built with numpy version A, then at run time it requires numpy version >=A. In this change, we read numpy version from the installed packages at build time, to avoid manually keeping the build time/runtime consistency.

Commit:
Update run_dockerbuild.sh: set default python version based on OS version (#8136)

hariharans29 committed:
Update logic in props.xml to account for shared provider library changes (#8138)

viboga committed:
Fuse attention node even in case of different Q,K hidden dimensions (#8106)

* changes to fuse attention node and create varied dimensions

* added an option to optimizer to only do offline fusion

* fixing a typo

* merge with master

* removing extra changes

* added new unit test - test_attention_fusion_for_varied_qkv_dimensions()

* Unit test succesfull for q,k,v paths with varied dimensions

* adding test model for unit test case

* optimizing attention tests

* removing debugs

* minor change

* addressing comments

* addressing comments

* changed the new option to disable_onnxruntime

* replacing asserts with debugs

* make attn fusion backward compatible for head_size, hidden_size

* preserving behavior for shape_modified_tensor

* adding new option as the last parameter

* cleaning up

* line breaks and spaces

* formatting according to python

* making the changes to fuse attention node without user input

* changes to fusion_attention.py updated

* bringing the code up to python standard

baijumeswani committed:
Interface for the module manager and implementation of the torch module manager

almaly-microsoft committed:
merge with bmeswani/module-wrapper

almaly-microsoft committed:
fixes for new module interface

almaly-microsoft committed:
fixes for new module interface

Commit:
test update

Commit:
Revert "test update"

This reverts commit dbd0c1090b34930e2edaba11f52d8f3592214e52.
